### DISCUSSION TITLE
Historical and future exogenous variables (like in Neuralforecast)

### DISCUSSION THREAD

**Discussion by tereli** [Nixtla/mlforecast#272] [General]
In NeuralForecast, exogenous variables can be labeled historical or future, depending on whether future values of the exogenous are known or not (https://nixtla.mintlify.app/neuralforecast/examples/exogenous_variables.html). 
As far as I understand, in MLForecast, all given exogenous features are automatically assumed to be future. Am I understanding correctly, and if so, is there any way to indicate a historic exogenous in MLForecast?

**Comment by jmoralez** [comment]
Hey. Yes, all exogenous features are assumed to be future. A possible approach to defining historic only would be providing a lagged version of the feature, e.g. if you're forecasting 20 periods ahead you can use the lag20 of your feature, that way you should have the required information at inference time. The transform_exog function may help with that, you can find the guide [here](https://nixtla.mintlify.app/mlforecast/docs/how-to-guides/transforming_exog.html).

Please let us know if you have further doubts.

**Reply by tereli** [reply]
Thank you! This helps me a lot :-)

**Reply by seanko** [reply]
there are lots model does not support future exogenous variables，such as ['PatchTST','iTransformer','SOFTS','TimesNet']，only fews support support future exogenous variables such as 'lstm'.'rnn','NHITS' ,is there a list make it clear ?if there had one ,let me know

**Comment by jmakov** [comment]
Can somebody elaborate why this is the default option instead of historical exos? Also doesn't seem it would be explicitly mentioned in the docs...

**Reply by jmoralez** [reply]
It's mentioned in the first paragraph of the [exogenous features guide](https://nixtlaverse.nixtla.io/mlforecast/docs/how-to-guides/exogenous_features.html)

> Keep in mind that all features in your input dataframe will be used for training, so you’ll have to provide the future values of exogenous features to [MLForecast.predict](https://nixtla.github.io/mlforecast/forecast.html#mlforecast.predict) through the X_df argument.

The reason is that this allows you to provide future values or "historic" values in the same way. You can find an example on how to create the required structure when you don't have the future values of your exogenous features [here](https://github.com/Nixtla/mlforecast/issues/328#issuecomment-2030009549).
